sai/features/features.py

```diff
@@ -0,0 +1,1589 @@
+
+    # Identify segregating sites (where 0 < p < 1)
+    segregating_mask = (gts_freq > 0) & (gts_freq < 1)
+
+    num_S = np.sum(segregating_mask)
+
+    if return_frequencies:
+        return num_S, gts_freq[segregating_mask]
+    return num_S
+
+
+def S_ABBA_BABA_calc(
+    pop1_freq: np.ndarray,
+    pop2_freq: np.ndarray,
+    pop3_freq: np.ndarray,
+    outgroup_freq: float = 0,
+    return_numerator_for_S: bool = False,
+) -> float:
+    """
+    Calculate the ABBA-BABA statistic, which is used in population genetics to test
+    for asymmetry in allele sharing between populations and an outgroup. The function
+    computes either the D-statistic (ABBA - BABA) / (ABBA + BABA) or the numerator
+    (ABBA - BABA) based on the specified parameters.
+
+    Parameters
+    ----------
+    pop1_freq : np.ndarray
+        An array of allele frequencies for population 1 at each locus. (Usually the reference population)
+
+    pop2_freq : np.ndarray
+        An array of allele frequencies for population 2 at each locus. (Usually the target population)
+
+    pop3_freq : np.ndarray
+        An array of allele frequencies for population 3 at each locus. (Usually the source population)
+
+    outgroup_freq : float, optional, default=0
+        The allele frequency in the outgroup population. The default is 0.
+
+    return_numerator_for_S : bool, optional, default=False
+        If True, the function returns the numerator (ABBA - BABA) instead of
+        the D-statistic. The default is False.
+
+    Returns
+    -------
+    float
+        The D-statistic (if `return_numerator_for_S` is False) or the numerator
+        (ABBA - BABA) if `return_numerator_for_S` is True. Returns `nan` if the
+        denominator is zero.
+
+    Notes
+    -----
+    The ABBA-BABA statistic is commonly used in studies of introgression or
+    gene flow between populations, where a high value of the D-statistic suggests
+    gene flow from population 1 to 2, and a low value suggests gene flow in the
+    opposite direction.
+    """
+
+    abbavec = (1.0 - pop1_freq) * pop2_freq * pop3_freq * (1 - outgroup_freq)
+    babavec = pop1_freq * (1.0 - pop2_freq) * pop3_freq * (1 - outgroup_freq)
+
+    # Summing up across loci
+    abba = np.sum(abbavec)
+    baba = np.sum(babavec)
+
+    if not return_numerator_for_S:
+        # Compute D-statistic
+        if (abba + baba) > 0:
+            return (abba - baba) / (abba + baba)
+        else:
+            return float("nan")
+    else:
+        return abba - baba
+
+
+def compute_ABBA_BABA_D(
+    src_gts: np.ndarray,
+    ref_gts: np.ndarray,
+    tgt_gts: np.ndarray,
+    out_gts: np.ndarray = None,
+    ploidy: int = 1,
+) -> float:
+    """
+    Computes Patterson's D-statistic (ABBA-BABA statistic) for detecting admixture between populations.
+
+    Parameters
+    ----------
+    src_gts : np.ndarray
+        A 2D numpy array representing haplotypes of population 1 (source).
```

Comment:
population 1 is the source population? However,

https://github.com/jalhackl/sai/blob/3c682640d9327f2779f1b6883ef9546ed7f04a74/sai/features/features.py#L556-L557

Reply:
and here

https://github.com/jalhackl/sai/blob/3c682640d9327f2779f1b6883ef9546ed7f04a74/sai/features/features.py#L666-L667

which one is correct?